rfd/0199-teleport-key-agent.md

```diff
@@ -0,0 +1,514 @@
+be entirely avoided, the same way that we can not stop a user from running
+`tsh scp $HOME/.tsh server01:`.
+
+As usual remote attacks are best mitigated through phish-proof authentication.
+If the following features are enabled, the impact of a compromised forwarded
+Teleport Key Agent will be largely mitigated:
+
+* Per-session MFA or Hardware Key Support (pin/touch)
+* MFA for Admin Actions
+
+If deemed necessary, we could put the agent behind an extra
+[confirmation layer](#confirmation-prompt-future-work). This confirmation layer
+would require dependent clients to perform a confirmation action (y/N, pin, or
+touch) to allow the dependent client to utilize the agent.
+
+### Configuration
+
+See [running the agent](#running-the-agent) for how to configure Teleport
+Connect to run the Teleport Key Agent.
+
+### Proto specification
+
+```proto
+// KeyAgentService provides a Teleport key agent service, allowing multiple Teleport client
+// processes to share a single instance of a Teleport key. This is useful for PIV keys which
+// can only be accessed by one process at a time.
+service KeyAgentService {
+  // Query checks whether the provided key index has an existing keyring and
+  // returns the public keys.
+  rpc Query(QueryRequest) returns (QueryResponse) {}
+  // Sign signs the given digest with the private key corresponding to the given
+  // key index and type. If a hash or salt was used to produce the digest, HashName
+  // and SaltLength must be provided as well.
+  //
+  // This rpc can be used to implement Go's crypto.Signer interface.
+  rpc Sign(SignRequest) returns (SignResponse) {}
+}
+
+// QueryRequest is a query request.
+message QueryRequest {
+  // KeyIndex is the index of the key.
+  KeyIndex key_index = 1;
+}
+
+// QueryResponse is a query response.
+message QueryResponse {
+  // SshPublicKey is the ssh public key.
+  bytes ssh_public_key = 1;
+  // TlsPublicKey is the tls public key.
+  bytes tls_public_key = 2;
+}
+
+// SignRequest is a sign request.
+message SignRequest {
+  // KeyIndex is the index of the key requested for signature.
+  KeyIndex key_index = 1;
+  // Digest is a hashed message to sign.
+  bytes digest = 2;
+  // HashName is the name of the hash used to generate the digest.
+  string hash_name = 3;
+  // SaltLength controls the length of the salt to use in PSS signature if set.
+  string salt_length = 4;
+}
+
+// SignResponse is a sign response.
+message SignResponse {
+  // signature is the resulting signature.
+  bytes signature = 1;
+}
+
+// KeyIndex is the index of the key.
+message KeyIndex {
+  // ProxyHost is the root proxy hostname that a key is associated with.
+  string proxy_host = 1;
+  // Username is the username that a key is associated with.
+  string username = 2;
+  // KeyType is the specific type of key, e.g. TLS or SSH.
+  // Required for Sign requests.
+  KeyType key_type = 3;
+}
+
+// KeyType is the type of key.
+enum KeyType {
```

Comment:
nit: This is not a key "type" but instead what the key is used for. KeyUsage maybe?